In-App Links in Channel's Bookmarks

toffguy77 opened the pull request:
# Add In-App Links to Channel Bookmarks

This PR adds support for in-app links in Channel Bookmarks using auto-detection based on URL scheme. No new bookmark type or UI checkbox needed - users simply paste `mattermost://` URLs and they work automatically.

## Summary

- **Single bookmark type**: Uses existing `link` type, no new `inapp_link` type
- **Auto-detection**: Client detects `mattermost://` scheme and handles as in-app navigation
- **Server-side security**: Validates URL schemes with whitelist + blacklist to prevent XSS
- **No UI changes**: No checkbox, no new fields - just paste the URL

## Implementation

### Server (Go)
- Added `IsValidLinkURL()` with scheme validation:
  - **Whitelist**: `http`, `https`, `mattermost`
  - **Blacklist**: `javascript`, `data`, `file`, `vbscript`, `about`
- Replaced `IsValidHTTPURL()` with `IsValidLinkURL()` in bookmark validation
- 22 tests covering security scenarios and edge cases

### Client (TypeScript)
- `isMattermostAppURL()`: Detects `mattermost://` scheme
- `parseMattermostLink()`: Converts `mattermost://` URLs to SPA paths
- Auto-detection in `bookmark_item.tsx`: Uses `history.push()` for seamless navigation
- 29 tests covering detection, parsing, and navigation

### Desktop
- Changes in https://github.com/mattermost/desktop/pull/3454
- Registers `mattermost://` protocol handler

### Mobile
- Changes in https://github.com/mattermost/mattermost-mobile/pull/8969
- Deep linking support

## Testing

- **Coverage**: 90% unit tests, 100% critical code
- **Total tests**: 51 (22 Go + 29 TypeScript)
- All security scenarios covered
- Backward compatible with existing bookmarks

## QA Test Steps

1. Open a channel and access the bookmarks menu
2. Create a new bookmark with a `mattermost://` URL (e.g., `mattermost://myteam/channels/town-square`)
3. Save the bookmark - it appears in the list like any other link
4. Click the bookmark - it navigates within the app without opening a new tab
5. Verify regular HTTP/HTTPS bookmarks still work as before
6. Try creating a bookmark with `javascript:alert('xss')` - should be rejected by server

## Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/29145

## Release Note

```release-note
Added support for in-app links in channel bookmarks. Users can now create bookmarks with mattermost:// URLs that navigate within the app instead of opening external browsers.
```

toffguy77 commented:
/check-cla

enahum commented on webapp/channels/src/components/channel_bookmarks/bookmark_item.tsx:
```diff
@@ -206,6 +220,21 @@ const DynamicLink = forwardRef<HTMLAnchorElement, DynamicLinkProps>(({
     const prefixed = href[0] === TARGET_BLANK_URL_PREFIX;
 
+    // Handle mattermost:// URLs
+    if (isInAppLink && href.startsWith('mattermost://')) {
```
Why hardcoded? This will not work for those who build their own apps

toffguy77 replied:
@enahum, thank you for your review and valuable comments!

I have addressed the issue with the hardcoded app scheme—now it is configurable. Could you please take another look?

Regarding the mobile app: as far as I know, its codebase is in a separate repository (mattermost/mattermost-mobile). My plan was to implement this feature there after finalizing the web/server-side changes.

Would you recommend creating a related PR in the mobile repo at this stage, or should I wait until this PR is reviewed and merged?

Thanks again for your feedback!

enahum replied:
I'll take another look sure.

Yes mobile is in another repo, but they should go together you don't want this to work on one and not the other

toffguy77 replied:
Mobile was implemented in https://github.com/mattermost/mattermost-mobile/pull/8969

enahum reviewed (changes requested):
Also, I don't see the implementation on the mobile app

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

enahum reviewed (changes requested):
What I see is that is can be set to any protocol and that means that not necessarily will open with MM, I wonder if we need this at all insteae of just supporting other protocols in the link option

toffguy77 committed:
Add support for in-app links in channel bookmarks

- Extend link bookmark validation to support mattermost:// scheme
- Add IsValidLinkURL() with whitelist (http, https, mattermost) and blacklist (javascript, data, file, vbscript, about)
- Auto-detect mattermost:// URLs on client and handle as in-app navigation
- No new bookmark type needed - single 'link' type handles all URL schemes
- No UI checkbox - automatic detection based on URL scheme

This allows users to create bookmarks with mattermost:// URLs that navigate
within the app instead of opening external browsers, while maintaining
security through scheme validation.

toffguy77 committed:
Add comprehensive tests for in-app links functionality

Server-side tests:
- Test mattermost:// URL validation
- Test dangerous schemes (javascript, data, file, vbscript, about)
- Test unsupported custom schemes rejection
- Add IsValidLinkURL unit tests

Client-side tests:
- Test isMattermostAppURL detection
- Test isUrlSafe with mattermost:// URLs
- Test isInternalURL recognition
- Test shouldOpenInNewTab behavior

toffguy77 committed:
Add missing isMattermostAppURL and parseMattermostLink functions

- Add isMattermostAppURL() to detect mattermost:// scheme
- Add parseMattermostLink() to parse mattermost:// URLs into SPA paths
- Support channel, permalink, DM, and global route patterns
- Add comprehensive tests for URL parsing logic

toffguy77 committed:
Fix Go code formatting (gofmt)

toffguy77 committed:
Add comprehensive tests for parseMattermostLink edge cases

- Test global routes (admin_console, oauth, etc.)
- Test generic fallback for custom paths
- Test team extraction from query parameter
- Test DM without @ (auto-add functionality)
- Test DM alias (dm vs messages)

Coverage: parseMattermostLink 60% → 100%
